src/psbt.c

```diff
@@ -2333,406 +2092,370 @@ int wally_psbt_sign(struct wally_psbt *psbt,
+        /* Make sure we don't already have a sig for this input */
+        size_t is_found;
+        ret = wally_map_find(&input->signatures, full_pubkey, full_pubkey_len, &is_found);
+        if (ret == WALLY_OK && !is_found)
+            ret = wally_map_find(&input->signatures, pubkey, pubkey_len, &is_found);
 
-        sighash_type = input->sighash_type ? input->sighash_type : WALLY_SIGHASH_ALL;
+        if (ret != WALLY_OK || is_found)
+            continue; /* Already got a partial sig for this pubkey on this input */
 
-        /* Get scriptcode and sighash */
-        if (input->redeem_script) {
-            unsigned char sh[WALLY_SCRIPTPUBKEY_P2SH_LEN];
-            size_t written;
+        /* From this point, any failure to sign returns an error, since we
+        * have the key to sign this input we are expected to be able to */
 
-            if ((ret = wally_scriptpubkey_p2sh_from_bytes(input->redeem_script, input->redeem_script_len, WALLY_SCRIPT_HASH160, sh, WALLY_SCRIPTPUBKEY_P2SH_LEN, &written)) != WALLY_OK) {
-                return ret;
-            }
-            if (input->non_witness_utxo) {
-                if (input->non_witness_utxo->outputs[txin->index].script_len != WALLY_SCRIPTPUBKEY_P2SH_LEN ||
-                    memcmp(sh, input->non_witness_utxo->outputs[txin->index].script, WALLY_SCRIPTPUBKEY_P2SH_LEN) != 0) {
-                    return WALLY_EINVAL;
-                }
-            } else if (input->witness_utxo) {
-                if (input->witness_utxo->script_len != WALLY_SCRIPTPUBKEY_P2SH_LEN ||
-                    memcmp(sh, input->witness_utxo->script, WALLY_SCRIPTPUBKEY_P2SH_LEN) != 0) {
-                    return WALLY_EINVAL;
-                }
-            } else {
-                continue;
-            }
-            scriptcode = input->redeem_script;
-            scriptcode_len = input->redeem_script_len;
-        } else {
-            if (input->non_witness_utxo) {
-                scriptcode = input->non_witness_utxo->outputs[txin->index].script;
-                scriptcode_len = input->non_witness_utxo->outputs[txin->index].script_len;
-            } else if (input->witness_utxo) {
-                scriptcode = input->witness_utxo->script;
-                scriptcode_len = input->witness_utxo->script_len;
-            } else {
-                continue;
-            }
-        }
+        if (!input_get_scriptcode(input, txin->index, &scriptcode, &scriptcode_len))
+            return WALLY_EINVAL; /* Couldn't find the script to sign with */
 
-        if (input->non_witness_utxo) {
-            unsigned char txid[WALLY_TXHASH_LEN];
+        sighash = input->sighash ? input->sighash : WALLY_SIGHASH_ALL;
 
-            if ((ret = wally_tx_get_txid(input->non_witness_utxo, txid, sizeof(txid))) != WALLY_OK) {
-                return ret;
-            }
-            if (memcmp(txid, txin->txhash, sizeof(txid)) != 0) {
-                return WALLY_EINVAL;
-            }
+        if (input->utxo) {
```

Comment:
Due to recent change in to hardware wallets, the full previous transaction is provided for segwit inputs https://github.com/bitcoin/bitcoin/pull/19215.

This means that we compute and sign the pre-segwit signature hash for segwit inputs, see https://github.com/ElementsProject/libwally-core/issues/213, https://github.com/bitcoin/bitcoin/issues/19650.

Reply:
Yes, this behaviour was buggy in the old code and remains so in the cleaned up code, I think this bug needs to be addressed in a new PR rather than getting caught up with the interface changes and general cleanup in this one.

Since we have https://github.com/ElementsProject/libwally-core/issues/213 to cover this issue I'll fix that bug separetly with a candidate PR we can hopefully get @Fonta1n3 to help us test.

Reply:
Absolutely, you guys ever need anything at all just tag me and I will get on it. Really appreciate all of that you do and would be happy to test help in anyway I can.